Switch from bower to webpack

nicksellen opened the pull request:
#### Proposed Changes

* switching from bower to webpack - [bower](https://bower.io/) says: _...psst! While Bower is maintained, we recommend using Yarn and Webpack or Parcel for front-end projects read how to migrate!_
* also prep for moving away from angular1 to react (see https://github.com/Trustroots/trustroots/issues/779#issuecomment-433696445 and https://github.com/Trustroots/trustroots/issues/638#issuecomment-415192512)

#### Testing Instructions

* run dev setup as normal
* ... which should run `npm install` and `npm run build` as part of it, if not, run those
* visit the site
* you should not notice anything changed (my favourite kind of contribution)

#### TODO

- [x] put bower in fire by integrating / rebasing with https://github.com/Trustroots/trustroots/pull/783
- [x] make tests work
- [x] build service worker scripts via webpack too
- [x] production build setup
- [x] check/fix profile networks (`Error: dst is undefined`)
- [x] check it works in docker
- [x] merge/adapt/rewrite dev scripts to use webpack build (and remove `build:webpack` npm script)

Some time later, or sooner, depending on motivation, but not essential for now, probably best in another PR:
- [ ] put gulp in fire
- [ ] style (css/less) via webpack
- [ ] add hot (module) reloading
- [ ] maybe use some cooler webpack options (splitting, etc) to make it suitable for production deployment
- [ ] reimplement `uib-templates.js` / `templates.js` building via webpack loader
- [ ] consider extracting and abstracting dependencies for legacy libs ([example from fs.de](https://gist.github.com/nicksellen/569a54762e9c2343cf27ffe15682a1aa))
- [ ] move away from massive lists of files in `config/assets/*`
- [ ] actually use js modules (currently all the actual `modules/*` files are written as if they are global) - can probably use some [webpack contexts](https://webpack.js.org/guides/dependency-management/) for globbing
- [ ] add es6 loader to webpack
- [ ] add at least 300 webpack plugins so computer can be used as heater in the coming winter

simison commented:
>maybe use some cooler webpack options (splitting, etc) to make it suitable for production deployment

We could leave most optimizations for another PR, especially since we don't declare any sort of dependency tree, so it makes sense to build just one bundle for now.

>style (css/less) via webpack

If it helps, this could easily be in another PR as well.

nicksellen commented:
> We could leave most optimizations for another PR, especially since we don't declare any sort of dependency tree, so it makes sense to build just one bundle for now.

You get most of it out-the-box with mode=production. The webpack peeps simplified it a lot since the old days. Development build cannot be used on production site, it's crazy and enormous.

simison commented:
Yap, I meant to leave fancier optimizations like code-splitting and tree-shaking for later tinkering. :thumbsup:

nicksellen commented:
This is rebased on top of #783 now :) ... and the errors from there fixed too.

simison commented on config/webpack/webpack.config.js:
```diff
@@ -0,0 +1,39 @@
+
+module.exports = merge(shims, {
+  mode: prod ? 'production' : 'development',
+  entry: require.resolve('./entries/main'),
```
Should we keep entry files as part of the app modules structure?

Something like `/modules/core/client/app/index.js` / `boot.js` / `main.js`

nicksellen replied:
That would then output one file per client module. I think it's better to leave webpack to split the files. What is common/easy with vue-router (and others I'm sure) is to use webpack dynamic loading so the resources for each page are loaded on demand without ever having to explicitly think about it.

The current approach here is basically a hack to leave the `assets/*` files as they are.

simison replied:
Gotcha, yeah Angular1 + ng-ui-router + Webpack should be popular enough combo that I'm sure someone has figured it out.

> The current approach here is basically a hack to leave the assets/* files as they are.

Makes sense! :thumbsup:

simison commented on config/webpack/webpack.config.js:
```diff
@@ -0,0 +1,39 @@
+const prod = process.env.NODE_ENV === 'production';
+
+module.exports = merge(shims, {
+  mode: prod ? 'production' : 'development',
```
nit: `prod` could be `isProduction` for clarity?

nicksellen replied:
Sure! Have a free :horse: too while I'm here :)

nicksellen commented:
I'm not too sure about the production build right now. `npm run start:prod` will bring up the homepage ok, but it seems to point to a different db and I didn't go further for now.

There is quite a lot of complexity and cruft in the scripts but I'd rather not address that in this PR. But there might be some improvements worth making.

Total bundled size is up a bit, from about 1.3mb (minified, no compression) to about 1.5mb. Seems not too bad.

nicksellen commented:
So, would be great to get it reviewed/tried out, I suspect there are a few things that can be made a bit smoother too, or clarified, etc.

nicksellen commented:
The profile network connections are broken, but don't know if that's just because it's local dev. Better go and prepare some food now :)

simison commented:
This is really neat work! :tada:

I've tested both development and production builds.

> I'm not too sure about the production build right now. npm run start:prod will bring up the homepage ok, but it seems to point to a different db and I didn't go further for now.

Yep, dev and prod use different collections via config. The actual deployment is built using `npm run build:prod`.

Production seems to build `push-messaging-sw.js` and other files nicely:

<img width="241" alt="image" src="https://user-images.githubusercontent.com/87168/47677392-a2bbed80-dbc7-11e8-890b-b4625338e5e1.png">


>There is quite a lot of complexity and cruft in the scripts but I'd rather not address that in this PR. >But there might be some improvements worth making.

Separate PRs better indeed. 👌 


>Total bundled size is up a bit, from about 1.3mb (minified, no compression) to about 1.5mb. Seems not too bad.

Yep, totally fine. 👍 New 0.44 mb and old 0.39 mb gzipped; at least we can optimize this much easier now! 💪 

>The profile network connections are broken, but don't know if that's just because it's local dev

Yeah they're not meant to work in `localhost` just like that. I connected via GitHub oAuth with some `/etc/hosts` juggling and it worked fine.


Great work!

I found one small issue with assets we could address before merging:

On profile edit page (http://localhost:3000/profile/edit), when modifying languages, the select directive is missing some images:

<img width="501" alt="image" src="https://user-images.githubusercontent.com/87168/47674412-e6125e00-dbbf-11e8-9d0b-ec97bf40dd44.png">

Here:
https://github.com/Trustroots/trustroots/blob/15a8bcebee8286eca1591f950a120420b2e08dd3/modules/core/client/less/modules/chosen.less#L6-L25

We could either use `file-loader` to load them from `node_modules` or alternatively just move them to [`/modules/core/client/img`](https://github.com/Trustroots/trustroots/tree/15a8bcebee8286eca1591f950a120420b2e08dd3/modules/core/client/img) since we've copied over the CSS from the module as well.


So looking forward to merge this!

simison commented:
> check it works in docker

FYI, I didn't test Docker setup yet.

Docker might need adding `bower` back?

nicksellen commented:
Ok, fixed up images (put them in `img/` folder as suggested, can't really use `file-loader` properly until we have the style going through webpack), and a couple of other minor things.

Didn't try docker, but should probably wait until someone has.

simison commented:
Greets from the Docker land:

```
> message-center@1.2.7 postinstall /trustroots/node_modules/message-center
> bower install

bower ESUDO         Cannot be run with sudo

Additional error details:
Since bower is a user command, there is no need to execute it with superuser permissions.
If you're having permission errors when using bower without sudo, please spend a few minutes learning more about how your system should work and make any necessary repairs.

http://www.joyent.com/blog/installing-node-and-npm
https://gist.github.com/isaacs/579814

You can however run a command with sudo using "--allow-root" option
npm WARN karma-webpack@3.0.5 requires a peer of webpack@^2.0.0 || ^3.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN ajv-keywords@3.2.0 requires a peer of ajv@^6.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN The package firebase is included as both a dev and production dependency.
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.4 (node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.4: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! message-center@1.2.7 postinstall: `bower install`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the message-center@1.2.7 postinstall script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2018-10-29T22_33_26_841Z-debug.log
ERROR: Service 'trustroots' failed to build: The command '/bin/sh -c npm install --quiet' returned a non-zero code: 1
```

What a bummer.

I just went ahead and forget that library and removed the `postinstall` script: https://github.com/Trustroots/message-center

Let's use that one instead.

nicksellen commented:
Oh the other thing to consider is the dev experience. At the moment I guess people change a js file and wait for livereload to reload browser or do it manually. 

I didn't want to put too much time into it here as would prefer to do it after gulp is gone with a proper hot reload setup. But it might be a worse experience until then. Maybe a minimal step is to at least run webpack in watch mode during dev.

nicksellen commented:
Ok, fixed a couple of things:
* switched to fork of message-centre, removed bower entirely
* fixed up webpack watch mode in dev (will rebuild and trigger browser reload)
* checked `npm run test:client:watch` works fine, it does

So, should be good to go from my perspective. Didn't run the docker setup.

simison commented:
Nice, I'll give this another spin later tonight. :thumbsup:

nicksellen commented:
Cool, just rebased on top of master again.

simison commented on .gitignore:
```diff
@@ -35,8 +34,9 @@ scipts/docker/nginx-conf/*
 
 # App
 build
-public/lib
```
Should we add this line back to ensure folks have a smooth transition?

This folder will exist with 1500 files and they'll appear in their `git status` the moment this goes to `master`.

simison commented:
Docker spins up now nicely! :thumbsup:

Live-reloading works smoothly as well.

I'll put back that previously ignored `public/lib` for a while so that other devs have it easier and merge.

🍻

simison commented:
Moved todo to https://github.com/Trustroots/trustroots/issues/791

mrkvon commented on .gitignore:
```diff
@@ -37,6 +36,8 @@ scipts/docker/nginx-conf/*
 build
 public/lib
```
`public/lib` can be removed

simison replied:
Yap, I put it back to be there for a little while so that simply pulling latest master wouldn't cause 1500 files to appear in `git status` when  `public/lib` disappears from gitignore.

Let's loop back to it and remove in a couple weeks.

Meanwhile, run `rm -fr public/lib` :-)
